packages/vite/src/node/config.ts

```diff
@@ -171,9 +176,16 @@ export interface UserConfig {
    */
   resolve?: ResolveOptions & { alias?: AliasOptions }
   /**
-   * CSS related options (preprocessors and CSS modules)
+   * CSS related options
+   * - CSSOptions is the default and provide configuration for PostCSS and
+   *   various preprocessors
+   * - LightningCSSOptions is an experimental option to handle CSS modules,
+   *   assets and imports via Lightning CSS. It requires to install it
+   *   as a peer dependency. This is incompatible with the use of preprocessors.
+   *
+   * See build options to configure CSS minification.
    */
-  css?: CSSOptions
+  css?: CSSOptions | LightningCSSOptions
```

Comment:
`CSSOptions` is:
```
export interface CSSOptions {
  /**
   * https://github.com/css-modules/postcss-modules
   */
  modules?: CSSModulesOptions | false
  preprocessorOptions?: Record<string, any>
  postcss?:
    | string
    | (PostCSS.ProcessOptions & {
        plugins?: PostCSS.AcceptedPlugin[]
      })
  /**
   * Enables css sourcemaps during dev
   * @default false
   * @experimental
   */
  devSourcemap?: boolean
}
```
Shouldn't we add a `lightningCSSOptions` to this interface instead?
I think we should still support `modules` and `devSourcemap`, no? And I agree with you that we can avoid preprocessor support, but it still makes sense to have the option available for later.

Reply:
In the case of LightningCSS:

- devSourceMap are on by default (I think this is fast enough)
- preprocessorOptions doesn't make sense
- modules?: doesn't have as much possibilities

So that why I opted for a complete new branch to let the two APIs evolve without restriction from the other

Reply:
I should check the whole PR before commenting. I see [the modules options are limited](https://github.com/parcel-bundler/lightningcss/blob/30469232c3fdb828753db4be49f48e9bf727a655/node/index.d.ts#L292) as you said
```
export interface CSSModulesConfig {
  /** The pattern to use when renaming class names and other identifiers. Default is `[hash]_[local]`. */
  pattern?: string,
  /** Whether to rename dashed identifiers, e.g. custom properties. */
  dashedIdents?: boolean
}
```

`css` was intended to be an object with Vite specific options though, that could be shared independently of the implementation.
I wonder if we should instead have
```
css: {
  transformer: 'lightingcss',
  lightningcss: {
    modules: ...,
    drafts: ...
  }
}
```
So equivalent to what we have now with `postcss` (we are also using with postcss, terser, the package name, so maybe it makes sense to switch to `lightningcss` instead of the capitalized version everywhere).

What we currently have in `modules`, we could start to add support for it in a way that works with Lightning later on if we find a way to do so. So these configs remain shared.

Reply:
I get the initial idea of "general css options independent of the implementation" but the reality is not here IMO:
 - modules is directly exposing the PostCSS possibilities
 - devSourceMap exist because there is some slowness in the implementation. As far I know, this option doesn't exist in the pipeline for example
 
And going in the "shared option" path will put burden on adding new features/implementation (example here for making all the combinaison of CSSModulesConfig available which I hope most people don't care, and if you care you should probably have better selector for e2e tests)

This is clearly a question of philosophy on how much should we abstract the underlining tools by default.

Reply:
Good points. It still feels a bit strange to override `css` depending on `transformer`. Do we have another example in our config? `css.modules` being a different type depending on `transformer` may make things more difficult to document/teach. So even if there aren't good examples of shared options (yet?), it could be a good idea to go with `css.lightningcss.modules`

Reply:
From my point of view this is simpler because you can document both branches separately. You don't need to tell "this doesn't work with that" (in the case of devSourceMap and css.modules)

I agree that if we would start now we could design a better API. Probably I'm biased by my background of working with GraphQL APIs, the notion of [discriminating unions](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions) feels very natural for me. But this is not in the new handbook anymore. 

This is something that I've seen in [zod](https://github.com/colinhacks/zod#discriminated-unions) and in [safeParse output](https://github.com/colinhacks/zod#safeparse) for example.

This makes the API less "indented". For me that ok that the configuration of CSS modules is not the same depending on the tool (but again this is really a question on how much we want to abstract underlying tools)

For the part of adding common option, we can add this in both branches if it happens to be the same, but the nice part of starting separate is that we can make this optional in one and not the other, have a different default, have a name that matches more the name of the underlying tool documentation, ...